crates/futures/src/queue.rs

```diff
@@ -1,31 +1,49 @@
 use js_sys::Promise;
 use std::cell::{Cell, RefCell};
 use std::collections::VecDeque;
+use std::mem;
 use std::rc::Rc;
 use wasm_bindgen::prelude::*;
 
 struct QueueState {
-    // The queue of Tasks which will be run in order. In practice this is all the
+    // The queue of Tasks which are currently run in order. In practice this is all the
     // synchronous work of futures, and each `Task` represents calling `poll` on
-    // a future "at the right time"
-    tasks: RefCell<VecDeque<Rc<crate::task::Task>>>,
+    // a future "at the right time". If we're not currently in run_all, this should be empty.
+    front_tasks: RefCell<VecDeque<Rc<crate::task::Task>>>,
+    // We use a double-buffer approach to running tasks. Tasks initially enter the
+    // back_tasks which gets swapped to the front when we resume spinning.
+    back_tasks: RefCell<VecDeque<Rc<crate::task::Task>>>,
```

Comment:
I think it's safe to always enqueue to a future tick because that's what JS promises do with `.then` and once you're using promises chances are that things will take awhile to complete and it'll be pretty rare to have an immediately enqueued piece of work anyway.

Reply:
Has been a bit more complicated to write a good test case, but I think I got it. The test case fails on `main` and succeeds with both approaches